README.rst
Python Client for OS Config API
===============================

|stable| |pypi| |versions|

`OS Config API`_: provides OS management tools that can be used for patch management, patch compliance, and configuration management on VM instances.

- `Client Library Documentation`_
- `Product Documentation`_

.. |stable| image:: https://img.shields.io/badge/support-stable-gold.svg
   :target: https://github.com/googleapis/google-cloud-python/blob/main/README.rst#stability-levels
.. |pypi| image:: https://img.shields.io/pypi/v/google-cloud-os-config.svg
   :target: https://pypi.org/project/google-cloud-os-config/
.. |versions| image:: https://img.shields.io/pypi/pyversions/google-cloud-os-config.svg
   :target: https://pypi.org/project/google-cloud-os-config/
.. _OS Config API: https://cloud.google.com/compute/docs/osconfig/
.. _Client Library Documentation: https://cloud.google.com/python/docs/reference/osconfig/latest
.. _Product Documentation:  https://cloud.google.com/compute/docs/osconfig/

Quick Start
-----------

In order to use this library, you first need to go through the following steps:

1. `Select or create a Cloud Platform project.`_
2. `Enable billing for your project.`_
3. `Enable the OS Config API.`_
4. `Setup Authentication.`_

.. _Select or create a Cloud Platform project.: https://console.cloud.google.com/project
.. _Enable billing for your project.: https://cloud.google.com/billing/docs/how-to/modify-project#enable_billing_for_a_project
.. _Enable the OS Config API.:  https://cloud.google.com/compute/docs/osconfig/
.. _Setup Authentication.: https://googleapis.dev/python/google-api-core/latest/auth.html

Installation
~~~~~~~~~~~~

Install this library in a `virtualenv`_ using pip. `virtualenv`_ is a tool to
create isolated Python environments. The basic problem it addresses is one of
dependencies and versions, and indirectly permissions.

With `virtualenv`_, it's possible to install this library without needing system
install permissions, and without clashing with the installed system
dependencies.

.. _`virtualenv`: https://virtualenv.pypa.io/en/latest/


Code samples and snippets
~~~~~~~~~~~~~~~~~~~~~~~~~

Code samples and snippets live in the `samples/` folder.


Supported Python Versions
^^^^^^^^^^^^^^^^^^^^^^^^^
Our client libraries are compatible with all current [active](https://devguide.python.org/devcycle/#in-development-main-branch) and [maintenance](https://devguide.python.org/devcycle/#maintenance-branches) versions of
Python.

Python >= 3.7

Unsupported Python Versions
^^^^^^^^^^^^^^^^^^^^^^^^^^^
Python <= 3.6

If you are using an [end-of-life](https://devguide.python.org/devcycle/#end-of-life-branches)
version of Python, we recommend that you update as soon as possible to an actively supported version.


Mac/Linux
^^^^^^^^^

.. code-block:: console

    pip install virtualenv
    virtualenv <your-env>
    source <your-env>/bin/activate
    <your-env>/bin/pip install google-cloud-os-config


Windows
^^^^^^^

.. code-block:: console

    pip install virtualenv
    virtualenv <your-env>
    <your-env>\Scripts\activate
    <your-env>\Scripts\pip.exe install google-cloud-os-config

Next Steps
~~~~~~~~~~

-  Read the `Client Library Documentation`_ for OS Config API
   to see other available methods on the client.
-  Read the `OS Config API Product documentation`_ to learn
   more about the product and see How-to Guides.
-  View this `README`_ to see the full list of Cloud
   APIs that we cover.

.. _OS Config API Product documentation:  https://cloud.google.com/compute/docs/osconfig/
.. _README: https://github.com/googleapis/google-cloud-python/blob/main/README.rst


CHANGELOG.md
# Changelog

## [1.15.3](https://github.com/googleapis/python-os-config/compare/v1.15.2...v1.15.3) (2023-09-13)


### Documentation

* Minor formatting ([1cebf49](https://github.com/googleapis/python-os-config/commit/1cebf49aa81cd0bf6994ce3ddc0eeccc2a50b65a))

## [1.15.2](https://github.com/googleapis/python-os-config/compare/v1.15.1...v1.15.2) (2023-07-04)


### Bug Fixes

* Add async context manager return types ([#280](https://github.com/googleapis/python-os-config/issues/280)) ([ba03471](https://github.com/googleapis/python-os-config/commit/ba0347119e6e14ea6a30793a1499ceaf60e46b56))

## [1.15.1](https://github.com/googleapis/python-os-config/compare/v1.15.0...v1.15.1) (2023-03-23)


### Documentation

* Fix formatting of request arg in docstring ([#272](https://github.com/googleapis/python-os-config/issues/272)) ([4fdfa8b](https://github.com/googleapis/python-os-config/commit/4fdfa8bb23c85bfb6785aa6e9450fdb042a3fb83))

## [1.15.0](https://github.com/googleapis/python-os-config/compare/v1.14.1...v1.15.0) (2023-02-21)


### Features

* Enable "rest" transport in Python for services supporting numeric enums ([#266](https://github.com/googleapis/python-os-config/issues/266)) ([679246f](https://github.com/googleapis/python-os-config/commit/679246ff80d7768e23fbdc7375515ecd6c1684b3))

## [1.14.1](https://github.com/googleapis/python-os-config/compare/v1.14.0...v1.14.1) (2023-01-20)


### Bug Fixes

* Add context manager return types ([97b46fa](https://github.com/googleapis/python-os-config/commit/97b46fa4d0bb76692202b9645d6489cc5e97b6b8))


### Documentation

* Add documentation for enums ([97b46fa](https://github.com/googleapis/python-os-config/commit/97b46fa4d0bb76692202b9645d6489cc5e97b6b8))

## [1.14.0](https://github.com/googleapis/python-os-config/compare/v1.13.0...v1.14.0) (2023-01-10)


### Features

* Add support for python 3.11 ([#257](https://github.com/googleapis/python-os-config/issues/257)) ([fa3a63f](https://github.com/googleapis/python-os-config/commit/fa3a63f72b1c1c6603655aae1c1457638f53b489))

## [1.13.0](https://github.com/googleapis/python-os-config/compare/v1.12.4...v1.13.0) (2022-12-13)


### Features

* Add support for `google.cloud.osconfig.__version__` ([0980ecc](https://github.com/googleapis/python-os-config/commit/0980eccc0b4d528719b13849f2ff9bb2707e4a21))
* Add typing to proto.Message based class attributes ([0980ecc](https://github.com/googleapis/python-os-config/commit/0980eccc0b4d528719b13849f2ff9bb2707e4a21))


### Bug Fixes

* Add dict typing for client_options ([0980ecc](https://github.com/googleapis/python-os-config/commit/0980eccc0b4d528719b13849f2ff9bb2707e4a21))
* **deps:** Require google-api-core &gt;=1.34.0, >=2.11.0  ([326bd8c](https://github.com/googleapis/python-os-config/commit/326bd8cc374f342187beb50f95308408fe839986))
* Drop usage of pkg_resources ([326bd8c](https://github.com/googleapis/python-os-config/commit/326bd8cc374f342187beb50f95308408fe839986))
* Fix timeout default values ([326bd8c](https://github.com/googleapis/python-os-config/commit/326bd8cc374f342187beb50f95308408fe839986))


### Documentation

* **samples:** Snippetgen handling of repeated enum field ([0980ecc](https://github.com/googleapis/python-os-config/commit/0980eccc0b4d528719b13849f2ff9bb2707e4a21))
* **samples:** Snippetgen should call await on the operation coroutine before calling result ([326bd8c](https://github.com/googleapis/python-os-config/commit/326bd8cc374f342187beb50f95308408fe839986))

## [1.12.4](https://github.com/googleapis/python-os-config/compare/v1.12.3...v1.12.4) (2022-10-07)


### Bug Fixes

* **deps:** Allow protobuf 3.19.5 ([#245](https://github.com/googleapis/python-os-config/issues/245)) ([adeb2d1](https://github.com/googleapis/python-os-config/commit/adeb2d1bc30f88c032a1079f9469046e6ca5b5f1))

## [1.12.3](https://github.com/googleapis/python-os-config/compare/v1.12.2...v1.12.3) (2022-09-29)


### Bug Fixes

* **deps:** Require protobuf >= 3.20.2 ([#243](https://github.com/googleapis/python-os-config/issues/243)) ([73c75aa](https://github.com/googleapis/python-os-config/commit/73c75aa2a64e2accb6093dd8b5011370fb02bfe4))

## [1.12.2](https://github.com/googleapis/python-os-config/compare/v1.12.1...v1.12.2) (2022-08-11)


### Bug Fixes

* **deps:** allow protobuf < 5.0.0 ([#227](https://github.com/googleapis/python-os-config/issues/227)) ([c3a2047](https://github.com/googleapis/python-os-config/commit/c3a2047ac58a718e6bc34b6f6f43f48e42b65f92))
* **deps:** require proto-plus >= 1.22.0 ([c3a2047](https://github.com/googleapis/python-os-config/commit/c3a2047ac58a718e6bc34b6f6f43f48e42b65f92))

## [1.12.1](https://github.com/googleapis/python-os-config/compare/v1.12.0...v1.12.1) (2022-07-13)


### Bug Fixes

* **deps:** require google-api-core>=1.32.0,>=2.8.0 ([#218](https://github.com/googleapis/python-os-config/issues/218)) ([50330f2](https://github.com/googleapis/python-os-config/commit/50330f24ebe9bcc6851725ad6669331a9349020b))

## [1.12.0](https://github.com/googleapis/python-os-config/compare/v1.11.2...v1.12.0) (2022-07-07)


### Features

* add audience parameter ([d700d11](https://github.com/googleapis/python-os-config/commit/d700d1171f2ceab7aaf362eeff385c9733cf456c))


### Bug Fixes

* **deps:** require google-api-core >= 2.8.0 ([#213](https://github.com/googleapis/python-os-config/issues/213)) ([d700d11](https://github.com/googleapis/python-os-config/commit/d700d1171f2ceab7aaf362eeff385c9733cf456c))
* require python 3.7+ ([#215](https://github.com/googleapis/python-os-config/issues/215)) ([397e04d](https://github.com/googleapis/python-os-config/commit/397e04d1b76dc8d5145109140ac41be84609adac))

## [1.11.2](https://github.com/googleapis/python-os-config/compare/v1.11.1...v1.11.2) (2022-06-03)


### Bug Fixes

* **deps:** require protobuf <4.0.0dev ([#205](https://github.com/googleapis/python-os-config/issues/205)) ([14b8400](https://github.com/googleapis/python-os-config/commit/14b8400e2f23158f402f18b4af7c5e1fac0b35c5))


### Documentation


owl-bot-staging/v1alpha/noxfile.py
# -*- coding: utf-8 -*-
# Copyright 2023 Google LLC
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#
import os
import pathlib
import shutil
import subprocess
import sys


import nox  # type: ignore

ALL_PYTHON = [
    "3.7",
    "3.8",
    "3.9",
    "3.10",
    "3.11",
]

CURRENT_DIRECTORY = pathlib.Path(__file__).parent.absolute()

LOWER_BOUND_CONSTRAINTS_FILE = CURRENT_DIRECTORY / "constraints.txt"
PACKAGE_NAME = subprocess.check_output([sys.executable, "setup.py", "--name"], encoding="utf-8")

BLACK_VERSION = "black==22.3.0"
BLACK_PATHS = ["docs", "google", "tests", "samples", "noxfile.py", "setup.py"]
DEFAULT_PYTHON_VERSION = "3.11"

nox.sessions = [
    "unit",
    "cover",
    "mypy",
    "check_lower_bounds"
    # exclude update_lower_bounds from default
    "docs",
    "blacken",
    "lint",
    "lint_setup_py",
]

@nox.session(python=ALL_PYTHON)
def unit(session):
    """Run the unit test suite."""

    session.install('coverage', 'pytest', 'pytest-cov', 'pytest-asyncio', 'asyncmock; python_version < "3.8"')
    session.install('-e', '.')

    session.run(
        'py.test',
        '--quiet',
        '--cov=google/cloud/osconfig_v1alpha/',
        '--cov=tests/',
        '--cov-config=.coveragerc',
        '--cov-report=term',
        '--cov-report=html',
        os.path.join('tests', 'unit', ''.join(session.posargs))
    )


@nox.session(python=DEFAULT_PYTHON_VERSION)
def cover(session):
    """Run the final coverage report.
    This outputs the coverage report aggregating coverage from the unit
    test runs (not system test runs), and then erases coverage data.
    """
    session.install("coverage", "pytest-cov")
    session.run("coverage", "report", "--show-missing", "--fail-under=100")

    session.run("coverage", "erase")


@nox.session(python=ALL_PYTHON)
def mypy(session):
    """Run the type checker."""
    session.install(
        'mypy',
        'types-requests',
        'types-protobuf'
    )
    session.install('.')
    session.run(
        'mypy',
        '--explicit-package-bases',
        'google',
    )


@nox.session
def update_lower_bounds(session):
    """Update lower bounds in constraints.txt to match setup.py"""
    session.install('google-cloud-testutils')
    session.install('.')

    session.run(
        'lower-bound-checker',
        'update',
        '--package-name',
        PACKAGE_NAME,
        '--constraints-file',
        str(LOWER_BOUND_CONSTRAINTS_FILE),
    )


@nox.session
def check_lower_bounds(session):
    """Check lower bounds in setup.py are reflected in constraints file"""
    session.install('google-cloud-testutils')
    session.install('.')

    session.run(
        'lower-bound-checker',
        'check',
        '--package-name',
        PACKAGE_NAME,
        '--constraints-file',
        str(LOWER_BOUND_CONSTRAINTS_FILE),
    )

@nox.session(python=DEFAULT_PYTHON_VERSION)
def docs(session):
    """Build the docs for this library."""

    session.install("-e", ".")
    session.install("sphinx==4.0.1", "alabaster", "recommonmark")

    shutil.rmtree(os.path.join("docs", "_build"), ignore_errors=True)
    session.run(
        "sphinx-build",
        "-W",  # warnings as errors
        "-T",  # show full traceback on exception
        "-N",  # no colors
        "-b",
        "html",
        "-d",
        os.path.join("docs", "_build", "doctrees", ""),
        os.path.join("docs", ""),
        os.path.join("docs", "_build", "html", ""),
    )


@nox.session(python=DEFAULT_PYTHON_VERSION)
def lint(session):
    """Run linters.

    Returns a failure if the linters find linting errors or sufficiently
    serious code quality issues.
    """
    session.install("flake8", BLACK_VERSION)
    session.run(
        "black",
        "--check",
        *BLACK_PATHS,
    )
    session.run("flake8", "google", "tests", "samples")


@nox.session(python=DEFAULT_PYTHON_VERSION)
def blacken(session):
    """Run black. Format code to uniform standard."""
    session.install(BLACK_VERSION)
    session.run(
        "black",
        *BLACK_PATHS,
    )


@nox.session(python=DEFAULT_PYTHON_VERSION)
def lint_setup_py(session):
    """Verify that setup.py is valid (including RST check)."""
    session.install("docutils", "pygments")
    session.run("python", "setup.py", "check", "--restructuredtext", "--strict")
